pkg/server/message.go

```diff
@@ -254,69 +263,150 @@ func (t *TranslateServiceServer) UpdateMessages(
+			if all[i].Messages[j].Status == model.MessageStatusUntranslated {
+				untranslatedMessagesLookup[all[i].Messages[j].ID] = &all[i].Messages[j]
+			}
 		}
 
 		// Create a new messages to store the messages that need to be translated
 		toBeTranslated := &model.Messages{
-			Language: messages.Language,
-			Original: messages.Original,
-			Messages: make([]model.Message, 0, len(originalMessages.Messages)),
+			Language: original.Language,
+			Messages: make([]model.Message, 0, len(untranslatedMessagesLookup)),
 		}
 
-		// Iterate over the original messages and add any missing messages to the toBeTranslated.Messages slice
-		for _, m := range originalMessages.Messages {
-			if _, ok := translatedMessageIDs[m.ID]; !ok {
-				toBeTranslated.Messages = append(toBeTranslated.Messages, m)
-			}
+		for _, msg := range untranslatedMessagesLookup {
+			toBeTranslated.Messages = append(toBeTranslated.Messages, *msg)
 		}
 
 		// Translate messages -
-		// untranslated text in toBeTranslated messages will be translated from original to target language.
-		targetLanguage := toBeTranslated.Language
-		toBeTranslated.Language = originalMessages.Language
+		// untranslated messages in toBeTranslated will be translated from original to target language.
+		targetLanguage := all[i].Language
 		translated, err := t.translator.Translate(ctx, toBeTranslated, targetLanguage)
 		if err != nil {
-			return status.Errorf(codes.Unknown, err.Error()) // TODO: For now we don't know the cause of the error.
+			return nil, fmt.Errorf("translator translate messages: %w", err)
 		}
 
-		// Append the translated messages to the existing messages
-		messages.Messages = append(messages.Messages, translated.Messages...)
-
-		// Save the updated translated messages
-		switch err := t.repo.SaveMessages(ctx, serviceID, &messages); {
-		default:
-			// noop
-		case errors.Is(err, repo.ErrNotFound):
-			return status.Errorf(codes.NotFound, "service not found")
-		case err != nil:
-			return status.Errorf(codes.Internal, "")
+		// Overwrite untranslated messages with translated messages
+		for _, translatedMessage := range translated.Messages {
+			if untranslatedMessage, ok := untranslatedMessagesLookup[translatedMessage.ID]; ok {
+				*untranslatedMessage = translatedMessage
+			}
 		}
 	}
 
-	return nil
+	return all, nil
 }
 
-// langExists returns true if the provided language exists in the provided model.Messages slice.
-func langExists(msgs []model.Messages, lang language.Tag) bool {
-	return slices.ContainsFunc(msgs, func(m model.Messages) bool {
-		return m.Language == lang
-	})
+/*
+getUntranslatedIDs returns a list of message IDs that have been altered e.g.
+ 1. The message.message has been changed
+ 2. The message with new message.ID has been added
+*/
+func getUntranslatedIDs(old, new *model.Messages) []string {
+	lookup := make(map[string]string, len(old.Messages))
+
+	for _, msg := range old.Messages {
+		lookup[msg.ID] = msg.Message
+	}
```

Comment:
Can we create a generic helper for a fast check of the list containing a value? Very common functionality.

Something like...

```go
type ContainsFn = func(T comparable) bool

func Contains[T comparable, S any](s []S, f func(S) T) ContainsFn {
    // prepare data for fast lookup

    return func(v T) bool {
        
    }
}
```

Reply:
We should be able to, but in separate PR, as this one is getting quite large